src/app/modules/service-providers/services/routes/routes.service.ts

```diff
@@ -13,11 +14,43 @@ export class RoutesService {
   constructor(@Inject(BINGPROVIDER_CONFIG) public bingProviderConfig: BingProviderInterface) { }
 
   getRoute(from, to): Observable<any> {
-    function indexLegs(rawLegs): Array<LegDetails> {
-      const indexedLegs = new Array<LegDetails>();
+    function mapLegs(rawLegs): Array<LegDetails> {
+      const mapedLegs = new Array<LegDetails>();
+      rawLegs.forEach(leg => {
+        const currentLeg: LegDetails = <LegDetails>{};
+        currentLeg.text = leg.instruction.text;
+        currentLeg.coords = leg.maneuverPoint.coordinates;
+        currentLeg.index = 0;
+        currentLeg.maneuverType = leg.details[0].maneuverType;
+        const legDetails = leg.details[0];
+        currentLeg.name = (legDetails.names && legDetails.names[0]) || (leg.signs && leg.signs[0]) || '???';
+        mapedLegs.push(currentLeg);
+      });
+      return mapedLegs;
+    }
+
+    function roundCoordinate(num: number) {
+      return Math.round(num * 10000) / 10000;
+    }
 
-      return indexedLegs;
+    function indexLegs(mapedLegs, linePoints ): Array<LegDetails> {
+      const line = lineString(linePoints);
+      mapedLegs.forEach(leg => {
+        const near = nearestPointOnLine(line, point(leg.coords));
+        leg.coords = near.geometry.coordinates;
+      });
+
+      let j = 0;
+      for (let i = 0; i < linePoints.length; i++) {
+        if (roundCoordinate(mapedLegs[j].coords[0]) === roundCoordinate(linePoints[i][0]) &&
+          roundCoordinate(mapedLegs[j].coords[1]) === roundCoordinate(linePoints[i][1])) {
+          mapedLegs[j].index = i;
+          j++;
+        }
+      }
+      return mapedLegs;
     }
+
     const resultSubject = new Subject<any>();
     const routeDetails: RouteDetails = new class implements RouteDetails {
```

Comment:
no need for new class, just object ({ ... }) with the required properties.
There is no need for this overhead.